### DISCUSSION TITLE
How to create a signature for ParameterizeSulcus2015.py

### DISCUSSION THREAD

**Discussion by mirkozaff** [brainvisa/brainvisa.github.io#64] [Q&A] [answered]
Hi,
I am trying to use ParameterizeSulcus2015.py as a standalone script to get sulcus information, I am using brainvisa 4.5 and I am getting this error when I try to define the Signature

` File "/home/zaffaro/brainvisa-4.5.0-mandriva/python/brainvisa/data/neuroDiskItems.py", line 2186, in getDiskItemType
    raise ValueError( HTMLMessage(_t_('<em><code>%s</code></em> is not a valid file or directory type') % str( item )) )
ValueError: <em><code>Cortical folds graph</code></em> is not a valid file or directory type`

Can anybody give me an example on how to correctly define a signature for the script?
Thank you!

**Comment by denisri** [comment, marked as answer]
Hi,
I don't completely understand what you are trying to do and how you are doing it...
Do you want to use an existing process (`ParameterizeSulcus2015`) in a script, or write a custom process ? In the first case you don't have to write a signature... (or maybe you mean fill it it from the script ?)
To run a process from a script, there are several options:
- from a shell script (as a commandline):
  ```
  python -m brainvisa.axon.runprocess ParameterizeSulcus2015 graph=/data/path_to_graph.arg # etc for other params
  ```
- as a python library from a python script:
  ```
  from brainvisa.axon import processes
  processes.initializeProcesses()
  from brainvisa.processes import defaultContext
  defaultContext().runProcess('ParameterizeSulcus2015', graph='/data/path_to_graph.arg') # + other params...
  ```

**Reply by mirkozaff** [reply]
Hello, I apologize for resuming this thread after some time, but now having had the opportunity to work better with brainvisa I have had time to understand clearly what my needs are.
I would like to know how to run the following command:

```
from brainvisa.axon import processes
processes.initializeProcesses()
from brainvisa.processes import defaultContext
defaultContext().runProcess('ParameterizeSulcus2015', graph='/data/path_to_graph.arg') # + other params...
```
Is it possible to just import brainvisa libraries in a python script? Do I have to lunch scripts from the "bv bash" command line everytime?

Thank you!

**Reply by Sofiia2001** [reply]
@mirkozaff Hello! Just came across your question and I was wondering if it worked for you to run brainvisa from python script? I am using the bv bash python environment and I cannot import anything on python 3 because it gives me weird errors of some "missing modules". When I import the processes module in python 2, it does not give an error, but throws one when I try to import the ParametrizeSulcus2015.py process, it says something like 

```
>>> processes.readProcess('ParameterizeSulcus2015.py')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/casa/install/python/brainvisa/processes.py", line 6168, in readProcess
    if _readProcessLog is not None:
NameError: global name '_readProcessLog' is not defined
```

Did you manage to setup python so that everything imports and works OK? 

Thank you in advance!

**Reply by denisri** [reply]
Hi,
The main BV distribution is a container, thus needs the `bv` indirection (no need to use bash, you can use `bv python` but anyway it's not in your native system).
The binary distributions 5.0.x are built for python2, not python3, which explains why it doesn't work for you with python3.
Before using the brainvisa API, it is necessary to initialize things as I have explained in the snippet above.
Then you should use `getProcessInstance()` rather than `readProcess()`:
```
from brainvisa.axon import processes
processes.initializeProcesses()
from brainvisa import processes
p = processes.getProcessInstance('ParameterizeSulcus2015')
```
If you need to run such a python script from outside the container, you thus can call it using `bv python this_script.py`. @mirkozaff: It's **not** possible to import brainvisa libraries from the host system outside of the container, if that was your question (which I forgot to answer long ago...), unless you rebuild all the environment and libraries on a host system, which is **complex** because of the numerous dependencies which need to be installed and complex libraries versions dependencies.
@Sofiia2001: to use python3, you need to rebuild everything from sources for python3. In a dedicated container, it's possible and fairly easy (just takes some time to build) using the "[developer environment](https://brainvisa.info/web/download.html#developer-environment-installation)" Casa-Distro containers, which already contain all the needed development tools and libraries dependencies. It will still be inside a container but will work for python3.

**Reply by Sofiia2001** [reply]
@denisri Thank you for your answer! I am not trying to run brainvisa from my local python, I'm using the python that is in singularity container, so I was assuming python3 would be set up already to work with bv modules there. But it seems that it is not configured properly. Also, I am reading the process from python file here because I provided some changes to it, so it is not a default one. That's why I thought I first have to identify this python script as a process type. 
I will try to run what you provided in your comment, maybe that will work. 
Thank you again!

**Reply by denisri** [reply]
If it's a personal or modified process, then you can save it (or make a symbolic link) in the `$HOME/.brainvisa/processes/` directory *of the container*. If this directory doesn't exist yet, you can create it. When Brainvisa starts again, it will consider it as a personal toolbox and processes there will be available. Take care however because the modified process with the same filename will override the one in the "regular" brainvisa...

**Comment by denisri** [comment]
Yes, that's what the python script lines in your quote above are doing, thus are the answer to your question. Did it work for you ?

**Reply by mirkozaff** [reply]
Ok thank you!
